commec/fetch_nc_bits.py

```diff
@@ -1,88 +1,128 @@
+    """
+    Get indices not covered by the query start / end ranges in the BLAST results.
+    """
+    unique_hits = blast_df.drop_duplicates(subset=["q. start", "q. end"])
+    hit_ranges = unique_hits[["q. start", "q. end"]].values.tolist()
 
+    # Sort each pair to ensure that start < end, then sort entire list of ranges
+    hit_ranges = sorted([sorted(pair) for pair in hit_ranges])
 
+    nc_ranges = []
+
+    # Include the start if the first hit begins more than 50 bp after the start
+    if hit_ranges[0][0] > 50:
+        nc_ranges.append([1, hit_ranges[0][0] - 1])
+
+    # Add ranges if there is a noncoding region of >=50 between hits
+    for i in range(len(hit_ranges) - 1):
+        nc_start = hit_ranges[i][1] + 1      # starts after this hit
+        nc_end = hit_ranges[i + 1][0] - 1    # ends before next hit
+
+        if nc_end - nc_start + 1 >= 50:
+            nc_ranges.append([nc_start, nc_end])
+
+    # Include the end if the last hit ends more than 50 bp before the end
+    query_length = blast_df['query length'][0]
+    if query_length - hit_ranges[-1][1] >= 50:
+        nc_ranges.append([hit_ranges[-1][1] + 1, int(query_length)])
+
+    return nc_ranges
+
+
+def get_records(fasta_file_path):
+    """Parse SeqIO records from input FASTA."""
+    with open(fasta_file_path, "r", encoding="utf-8") as fasta_file:
+        records = list(SeqIO.parse(fasta_file, "fasta"))
+        return records
+
+
+def write_nc_sequences(nc_ranges, record, outfile: str):
+    """
+    Write a FASTA containing only the parts of the record without a high-quality protein match.
+    """
+    nc_sequences = []
+    for start, stop in nc_ranges:
+        # subtract 1 just from `start` to adjust to 0-based index and capture full range
+        sequence = record.seq[int(start) - 1 : int(stop)]
+        nc_sequences.append(f">{record.id} {record.description} {start}-{stop}\n{sequence}\n")
+
+    with open(outfile, "w", encoding="utf-8") as output_file:
+        output_file.writelines(nc_sequences)
+
+
+def main(protein_results, query_fasta):
+    """Fetch noncoding regions > 50bp and write to a new file."""
+    outfile = re.sub(".nr.*", "", protein_results) + ".noncoding.fasta"
+
+    sys.stdout.write(f"\t...checking protein hits in: {protein_results}\n")
+
+    if is_empty(protein_results) or not has_hits(protein_results):
+        sys.stdout.write("\t...no protein hits found, screening entire sequence\n")
+        shutil.copyfile(query_fasta, outfile)
+        return
+
+    blast_df = get_high_identity_matches(protein_results)
+
+    if blast_df.empty:
+        sys.stdout.write(
+            "\t...protein hits all low percent identity (<90%) - screening entire sequence\n"
+        )
+        shutil.copyfile(query_fasta, outfile)
+        return
+
+    sys.stdout.write(
+        "\t...protein hits found, fetching nt regions not covered by a 90% ID hit or better\n"
+    )
+    ranges_to_screen = get_ranges_with_no_hits(blast_df)
```

Comment:
To follow up on this, one of the issues I have found in my json outputs is exactly this, matching blast query accessions to whats in our fasta - as there is parsing that often occurs similar to what SeqIO does too I think? In short however, it does seem like comparing the first 25 characters instead of the whole string may prove a very simple work-around to all these internal query ID comparison issues.